homeassistant/components/xs1.py

```diff
@@ -0,0 +1,117 @@
+"""
+Support for the EZcontrol XS1 gateway.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/xs1/
+"""
+
+import asyncio
+import logging
+from functools import partial
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_HOST, \
+    CONF_PORT, CONF_USERNAME, CONF_PASSWORD
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['xs1-api-client==2.3.5']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'xs1'
+ACTUATORS = 'actuators'
+SENSORS = 'sensors'
+
+# configuration keys
+CONF_SSL = 'ssl'
+
+# define configuration parameters
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT): cv.string,
+        vol.Optional(CONF_SSL): cv.boolean,
+        vol.Optional(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD): cv.string
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+XS1_COMPONENTS = [
+    'switch',
+    'sensor',
+    'climate'
+]
+
+# Lock used to limit the amount of concurrent update requests
+# as the XS1 Gateway can only handle a very
+# small amount of concurrent requests
+UPDATE_LOCK = asyncio.Lock()
+
+
+def _create_controller_api(host, port, ssl, user, password):
+    import xs1_api_client
+    return xs1_api_client.XS1(
+        host=host,
+        port=port,
+        ssl=ssl,
+        user=user,
+        password=password)
+
+
+async def async_setup(hass, config):
+    """Set up XS1 Component."""
+    _LOGGER.debug("Initializing XS1")
+
+    host = config[DOMAIN].get(CONF_HOST)
+    port = config[DOMAIN].get(CONF_PORT)
+    ssl = config[DOMAIN].get(CONF_SSL, False)
+    user = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+
+    # initialize XS1 API
+
+    xs1 = await hass.async_add_executor_job(
+        partial(_create_controller_api,
+                host, port, ssl, user, password))
```

Comment:
You should check the credentials. If they are not valid then the component setup should fail with a log entry.

Also, it could be that the `host` is offline. This could be handled with `PlatformNotReady`.

Reply:
I'm not exactly sure how easy it is to check credentials, I never used them on my device. I will try to check on that later.

Reply:
Ok so I checked again with my gateway and I simply cannot use the password feature (seems like a firmware bug which prevents me from setting a password). I guess that's why I never used authentication. I don't think this is a severe issue though as I don't think many people still use this gateway and it can be isolated in the network pretty easily as it doesn't need to communicate with anything else other than home assistant over the network.

If someone else has a gateway with a non default password I am happy to include code that validates credentials on initialization but I just can't test this atm :'(